fnet/data/bufferedpatchdataset.py

```diff
@@ -1,110 +1,180 @@
+        self.patch_shape = patch_shape
         self.buffer_size = min(len(self.dataset), buffer_size)
-        self.buffer_switch_frequency = buffer_switch_frequency
-        self.npatches = npatches
-        self.buffer = list()
-        self.verbose = verbose
+        self.buffer_switch_interval = buffer_switch_interval
         self.shuffle_images = shuffle_images
-        self.dim_squeeze = dim_squeeze
-        shuffed_data_order = np.arange(0, len(dataset))
-        if self.shuffle_images:
-            np.random.shuffle(shuffed_data_order)
-        pbar = tqdm(range(0, self.buffer_size))
-        self.buffer_history = list()
-        for i in pbar:
-            # convert from a torch.Size object to a list
-            if self.verbose:
-                pbar.set_description("buffering images")
-            datum_index = shuffed_data_order[i]
-            datum = dataset[datum_index]
-            datum_size = datum[0].size()
-            self.buffer_history.append(datum_index)
-            self.buffer.append(datum)
-        self.remaining_to_be_in_buffer = shuffed_data_order[i + 1:]
-        assert len(patch_size) + 1 == len(datum_size)
-        self.patch_size = [datum_size[0]] + list(patch_size)
-
-    def __len__(self):
-        return self.npatches
-
-    def __getitem__(self, index):
+
+        self.counter = 0
+        self.epochs = -1  # incremented to 0 when buffer initially filled
+        self.buffer = deque()
+        self.remaining_to_be_in_buffer = deque()
+        self.buffer_history = []
+        for _ in tqdm(range(self.buffer_size), desc='Buffering images'):
+            self.insert_new_element_into_buffer()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        patch = self.get_random_patch()
         self.counter += 1
-        if (self.buffer_switch_frequency > 0) and (
-            self.counter % self.buffer_switch_frequency == 0
+        if (self.buffer_switch_interval > 0) and (
+                self.counter % self.buffer_switch_interval == 0
         ):
-            if self.verbose:
-                logger.info("Inserting new item into buffer")
             self.insert_new_element_into_buffer()
-        return self.get_random_patch()
-
-    def insert_new_element_into_buffer(self):
-        # sample with replacement
-        self.buffer.pop(0)
-        if self.shuffle_images:
+        return patch
 
-            if len(self.remaining_to_be_in_buffer) == 0:
-                self.remaining_to_be_in_buffer = np.arange(
-                    0, len(self.dataset)
+    def _check_last_datum(self) -> None:
+        """Checks last dataset item added to buffer."""
+        nd = len(self.patch_shape)
+        idx_buf = self.buffer_history[-1]
+        shape_spatial = None
+        for idx_c, component in enumerate(self.buffer[-1]):
+            if shape_spatial is None:
+                shape_spatial = component.shape[-nd:]
+            elif component.shape[-nd:] != shape_spatial:
```

Comment:
This block of code is a little hard to follow. Particularly the negative indexes in the slice notation. A comment or two would make this a lot more transparent.